File: boj/bfs/7576.swift
import Foundation

struct Position {
    var y: Int, x: Int, time: Int
    
    init(_ y: Int, _ x: Int, _ time: Int) {
        self.y = y
        self.x = x
        self.time = time
    }
}

let toward = [[0, 1], [0, -1], [1, 0], [-1, 0]]

func bfs() -> Bool {
    var que = [Position]()
    var visited = [[Bool]](repeating: [Bool](repeating: false, count: N), count: M)
    
    for tomato in tomatos {
        que.append(tomato)
        visited[tomato.y][tomato.x] = true
    }
    
    var head = 0
    while que.count > head {
        let cur = que[head] // que에서 removeFirst() 하는 방식으로는 시간초과가 난다. head를 이용!
        head+=1
        empty-=1
        
        if empty == 0 {
            print("\(cur.time)")
            return true
        }
        
        for i in 0..<4 {
            let nextY = cur.y + toward[i][0]
            let nextX = cur.x + toward[i][1]
            
            if nextX < 0 || nextY < 0 || nextX >= N || nextY >= M { continue }
            if visited[nextY][nextX] { continue }
            if map[nextY][nextX] == -1 { continue }
            
            visited[nextY][nextX] = true
            que.append(Position(nextY, nextX, cur.time+1))
        }
    }
    
    return false
}

guard let input = readLine()?.split(separator: " ").compactMap({Int(String($0))}) else { exit(0) }

var N = input[0], M = input[1]

var map = [[Int]](repeating: [Int](repeating: 0, count: N), count: M)
var tomatos = [Position]()

var empty = N*M

for i in 0..<M {
    guard let row = readLine()?.split(separator: " ").compactMap({Int(String($0))}) else { exit(0) }
    for j in 0..<N {
        map[i][j] = row[j]
        if map[i][j] == 1 {
            tomatos.append(Position(i, j, 0))
        }
        if map[i][j] == -1 { empty -= 1 }
    }
}

if empty == 0 { print("0") } // 전체가 -1인 경우가 있기 때문에
if !bfs() { print("-1") }

File: boj/bfs/7576.java
import java.util.*;
import java.io.*;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    static int N, M;
    static int[][] map;
    static int[] move = {0, 0, 1, -1};
    static ArrayList<Pair> ripeList;
    static int unripeCnt;

    static class Pair {
        int x, y, count;

        Pair(int x, int y) {
            this.x = x;
            this.y = y;
        }

        Pair(int x, int y, int count) {
            this.x = x;
            this.y = y;
            this.count = count;
        }
    }

    static int bfs() {
        Queue<Pair> queue = new LinkedList<>();
        boolean[][] visited = new boolean[N][M];

        for (int i = 0; i < ripeList.size(); i++) {
            queue.add(new Pair(ripeList.get(i).x, ripeList.get(i).y, 0));
            visited[ripeList.get(i).x][ripeList.get(i).y] = true;
        }

        while (!queue.isEmpty()) {
            Pair cur = queue.poll();
            if (map[cur.x][cur.y] == 0) unripeCnt--;

            if (unripeCnt == 0) {
                return cur.count;
            }

            for (int i = 0; i < 4; i++) {
                int nextX = cur.x + move[i];
                int nextY = cur.y + move[3 - i];

                if (nextX < 0 || nextY < 0 || nextX >= N || nextY >= M) continue;
                if (visited[nextX][nextY]) continue;

                if (map[nextX][nextY] == 0) {
                    visited[nextX][nextY] = true;
                    queue.add(new Pair(nextX, nextY, cur.count + 1));
                }

            }
        }

        return -1;
    }

    public static void main(String[] args) throws IOException {
        String[] NM = br.readLine().split(" ");

        M = Integer.parseInt(NM[0]);
        N = Integer.parseInt(NM[1]);

        map = new int[N][M];

        ripeList = new ArrayList<>();
        unripeCnt = 0;
        for (int i = 0; i < N; i++) {
            String[] input = br.readLine().split(" ");
            for (int j = 0; j < M; j++) {
                map[i][j] = Integer.parseInt(input[j]);
                if (map[i][j] == 1) {
                    ripeList.add(new Pair(i, j));
                }
                else if (map[i][j] == 0) {
                    unripeCnt++;
                }
            }
        }

        int answer = bfs();
        System.out.println(answer);

    }
}